### ISSUE TITLE
ENH Add reweighing preprocessing technique based on Kamiran, Calders

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#784]
The paper [Data preprocessing techniques for classification
without discrimination](https://link.springer.com/content/pdf/10.1007%2Fs10115-011-0463-8.pdf) by Kamiran and Calders outlines preprocessing techniques *suppression*, *massaging*, and *reweighing* or *resampling*. The goal of this task is to implement the *reweighing* technique for `fairlearn.preprocessing`.

Even though this technique is from 2012 it is still frequently used as a baseline to compare to, so Fairlearn should provide it to users as well.

Completing this item requires:
- preprocessing technique code in `fairlearn.preprocessing`
- unit tests in `test.unit.preprocessing`
- descriptive API reference (directly in the docstring)
- ideally a short user guide in `docs.user_guide.mitigation.rst`

A fully fledged example notebook is not required.

To claim this task please respond below. Of course, you can also ask questions!

**Comment by galuhsahid** [comment]
Hello! I would like to work on this if that's okay. Will ask further questions once I start looking into it.

**Comment by romanlutz** [comment]
@galuhsahid Certainly! Welcome to Fairlearn! As you probably noticed the issue description isn't very detailed, so if you think it's helpful you can outline a plan for the implementation in this issue as well. That's entirely up to you, though. We're happy to review PRs, too. If you'd like to discuss anything further outside this issue we have a [Gitter](https://gitter.im/fairlearn/community) channel. Don't hesitate to ask questions :-)

**Comment by romanlutz** [comment]
@galuhsahid I just wanted to check in how it's going. There's no particular time pressure, but if you need help let us know. Also, if you're not planning to continue we can make this available to someone else. We're also starting to do pair programming sometimes, so if that's interesting to you please don't hesitate to ask.

**Comment by galuhsahid** [comment]
@romanlutz Hello! Sorry for the delay, I'm currently working on the implementation plan and will reach out to ask for feedback before going ahead with the code.

**Comment by romanlutz** [comment]
No problem at all! Reach out if you want to discuss anything :-)

**Comment by galuhsahid** [comment]
Hello! I would like to confirm a few things regarding the implementation:

1. Will the input type be pandas DataFrame?
2. For the output, should we return an array of the new weights?

**Comment by romanlutz** [comment]
We typically accept a variety of input types. See the `fit` method of `ThresholdOptimizer`, for example: https://fairlearn.org/main/api_reference/fairlearn.postprocessing.html#fairlearn.postprocessing.ThresholdOptimizer.fit

![image](https://user-images.githubusercontent.com/10245648/122808989-6363d300-d282-11eb-9952-ba100e73b3c3.png)

In terms of API I would imagine that we have
```
class KamiranCaldersReweighing(...):
    def __init__(self, ...):
        ...
    def fit(X, y, *, sensitive_features):
        ...
```
After calling `fit` the weights would be stored on the object (let's say in a member called `weights`, so we could either retrieve it using 
```
reweighing_obj = KamiranCaldersReweighing(...)
reweighing_obj.fit(X, y, sensitive_features=sensitive_features)
w = reweighing_obj.weights
```
or perhaps with a method `get_weights()` but it feels odd to write a custom getter like that. @adrinjalali @MiroDudik may have thoughts on this. I checked the scikit-learn docs for conventions but didn't really find anything similar (preprocessing that results in weights). Regardless of how we pass it on to the estimator we use for training we definitely need to create the weights internally and store them on the object.

I imagine all of this would go into the `fairlearn.preprocessing` module.

I'm still unsure about naming - can we come up with something better than what I put above using the author names? Or are we fine with that?

**Comment by adrinjalali** [comment]
Reweighing is gonna be tricky to have scikit-learn compatible until we have the "resamplers" on the scikit-learn side. For more info you can have a look here: https://github.com/scikit-learn/enhancement_proposals/pull/15

**Comment by hildeweerts** [comment]
### re. API
If I recall correctly the implementation in AIF360 does something similar to what @romanlutz proposes. Quite a few of my students had issues with that, because it wasn't clear to them that they should retrieve the weights and pass them separately to the estimator. I can imagine we could add an (optional?) estimator argument in which case it would be used more like a wrapper than a separate pre-processing module, but I can also see how that could open another can of worms... @adrinjalali do you expect this is something that will be taken up soon from the scikit-learn side? 

### re. naming
I don't like it a ton, but given that fair ml researchers haven't been very productive in coming up with unique names, I do think going with the (first) author makes the most sense. If we want to adopt this we should probably also update our older classes though...

**Comment by adrinjalali** [comment]
I wouldn't necessarily wait for hit @hildeweerts , but we also don't really have to. The meta estimator alternative mentioned [here](https://github.com/scikit-learn/enhancement_proposals/blob/c64044de8332b070eae9c7049a027133952d17eb/slep005/proposal.rst) which also has a prototype implementation on this PR:https://github.com/scikit-learn/scikit-learn/pull/13269 would be something we could already do.

**Comment by galuhsahid** [comment]
Hi all, thank you for the answers! If I can summarize so far:
- the input will accept various types, namely: np.ndarray, pd.DataFrame, pd.Series, list depending on the parameters
- as for the output, the weights will be stored on the object
- the code will go to the `fairlearn.preprocessing` module

Shall we proceed with implementing the meta estimator alternative? I’m assuming this will also be similar to https://github.com/fairlearn/fairlearn/blob/main/fairlearn/postprocessing/_threshold_optimizer.py where it takes an estimator argument, am I understanding it correctly? Thank you!

**Comment by hildeweerts** [comment]
Hi, @galuhsahid! Do you still intend to work on this? No worries if not!

If I receive no response in 2 weeks (before January 27) I will reassign the issue so other folks who are interested can work on it.

**Comment by SeanMcCarren** [comment]
I made some progress on this :)
